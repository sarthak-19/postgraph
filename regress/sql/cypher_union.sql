/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

LOAD 'postgraph';
SET search_path TO postgraph;

SELECT create_graph('cypher_union');

SELECT * FROM cypher('cypher_union', $$CREATE ()$$) as (a gtype);


SELECT * FROM cypher('cypher_union', $$MATCH (n) RETURN n UNION MATCH (n) RETURN n$$) as (a vertex);

SELECT * FROM cypher('cypher_union', $$MATCH (n) RETURN n UNION ALL MATCH (n) RETURN n$$) as (a vertex);

SELECT * FROM cypher('cypher_union', $$MATCH (n) RETURN n UNION RETURN 1$$) as (a vertex);

SELECT * FROM cypher('cypher_union', $$MATCH (n) RETURN n UNION RETURN NULL$$) as (a vertex);

SELECT * FROM cypher('cypher_union', $$RETURN [1,2,3] UNION RETURN 1$$) as (a gtype);

/*should return 1 row*/
SELECT * FROM cypher('cypher_union', $$RETURN NULL UNION RETURN NULL$$) AS (result gtype);

/*should return 2 rows*/
SELECT * FROM cypher('cypher_union', $$RETURN NULL UNION ALL RETURN NULL$$) AS (result gtype);

/*
 *multiple unions, precedence
 */
SELECT * FROM cypher('cypher_union', $$MATCH (n) RETURN n UNION MATCH (n) RETURN n UNION MATCH (n) RETURN n$$) AS (result vertex);

/*should return triple*/
SELECT * FROM cypher('cypher_union', $$MATCH (n) RETURN n UNION ALL MATCH (n) RETURN n UNION ALL MATCH(n) RETURN n$$) AS (result vertex);

/*should return single*/
SELECT * FROM cypher('cypher_union', $$MATCH (n) RETURN n UNION ALL MATCH (n) RETURN n UNION MATCH(n) RETURN n$$) AS (result vertex);

/*should return just a pair*/
SELECT * FROM cypher('cypher_union', $$MATCH (n) RETURN n UNION MATCH (n) RETURN n UNION ALL MATCH(n) RETURN n$$) AS (result vertex);

/*should return 3 null rows*/
SELECT * FROM cypher('cypher_union', $$RETURN NULL UNION ALL RETURN NULL UNION ALL RETURN NULL$$) AS (result gtype);

/*should return 1 null row*/
SELECT * FROM cypher('cypher_union', $$RETURN NULL UNION ALL RETURN NULL UNION RETURN NULL$$) AS (result gtype);

/*should return 2 null rows*/
SELECT * FROM cypher('cypher_union', $$RETURN NULL UNION RETURN NULL UNION ALL RETURN NULL$$) AS (result gtype);

/* scoping */
SELECT * FROM cypher('cypher_union', $$MATCH (n) RETURN n UNION ALL MATCH (m) RETURN n$$) AS (result vertex);

/*
 *UNION and UNION ALL, type casting
 */
SELECT * FROM cypher('cypher_union', $$RETURN 1.0::int UNION RETURN 1::float UNION ALL RETURN 2.0::float$$) AS (result gtype);

SELECT * FROM cypher('cypher_union', $$RETURN 1.0::int UNION RETURN 1.0::float UNION ALL RETURN 1::int$$) AS (result gtype);

SELECT * FROM cypher('cypher_union', $$RETURN 1.0::float UNION RETURN 1::int UNION RETURN 1::float$$) AS (result gtype);


SELECT drop_graph('cypher_union', true);
