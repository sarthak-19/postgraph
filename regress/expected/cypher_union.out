/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */
LOAD 'postgraph';
SET search_path TO postgraph;
SELECT create_graph('cypher_union');
NOTICE:  graph "cypher_union" has been created
 create_graph 
--------------
 
(1 row)

SELECT * FROM cypher('cypher_union', $$CREATE ()$$) as (a gtype);
 a 
---
(0 rows)

SELECT * FROM cypher('cypher_union', $$MATCH (n) RETURN n UNION MATCH (n) RETURN n$$) as (a vertex);
ERROR:  could not identify an equality operator for type vertex
LINE 1: ... * FROM cypher('cypher_union', $$MATCH (n) RETURN n UNION MA...
                                                             ^
SELECT * FROM cypher('cypher_union', $$MATCH (n) RETURN n UNION ALL MATCH (n) RETURN n$$) as (a vertex);
                           a                            
--------------------------------------------------------
 {"id": 281474976710657, "label": "", "properties": {}}
 {"id": 281474976710657, "label": "", "properties": {}}
(2 rows)

SELECT * FROM cypher('cypher_union', $$MATCH (n) RETURN n UNION RETURN 1$$) as (a vertex);
ERROR:  UNION could not convert type gtype to vertex
LINE 1: ...'cypher_union', $$MATCH (n) RETURN n UNION RETURN 1$$) as (a...
                                                             ^
SELECT * FROM cypher('cypher_union', $$MATCH (n) RETURN n UNION RETURN NULL$$) as (a vertex);
ERROR:  UNION could not convert type gtype to vertex
LINE 1: ...'cypher_union', $$MATCH (n) RETURN n UNION RETURN NULL$$) as...
                                                             ^
SELECT * FROM cypher('cypher_union', $$RETURN [1,2,3] UNION RETURN 1$$) as (a gtype);
     a     
-----------
 [1, 2, 3]
 1
(2 rows)

/*should return 1 row*/
SELECT * FROM cypher('cypher_union', $$RETURN NULL UNION RETURN NULL$$) AS (result gtype);
 result 
--------
 
(1 row)

/*should return 2 rows*/
SELECT * FROM cypher('cypher_union', $$RETURN NULL UNION ALL RETURN NULL$$) AS (result gtype);
 result 
--------
 
 
(2 rows)

/*
 *multiple unions, precedence
 */
SELECT * FROM cypher('cypher_union', $$MATCH (n) RETURN n UNION MATCH (n) RETURN n UNION MATCH (n) RETURN n$$) AS (result vertex);
ERROR:  could not identify an equality operator for type vertex
LINE 4: ... * FROM cypher('cypher_union', $$MATCH (n) RETURN n UNION MA...
                                                             ^
/*should return triple*/
SELECT * FROM cypher('cypher_union', $$MATCH (n) RETURN n UNION ALL MATCH (n) RETURN n UNION ALL MATCH(n) RETURN n$$) AS (result vertex);
                         result                         
--------------------------------------------------------
 {"id": 281474976710657, "label": "", "properties": {}}
 {"id": 281474976710657, "label": "", "properties": {}}
 {"id": 281474976710657, "label": "", "properties": {}}
(3 rows)

/*should return single*/
SELECT * FROM cypher('cypher_union', $$MATCH (n) RETURN n UNION ALL MATCH (n) RETURN n UNION MATCH(n) RETURN n$$) AS (result vertex);
ERROR:  could not identify an equality operator for type vertex
LINE 2: ... * FROM cypher('cypher_union', $$MATCH (n) RETURN n UNION AL...
                                                             ^
/*should return just a pair*/
SELECT * FROM cypher('cypher_union', $$MATCH (n) RETURN n UNION MATCH (n) RETURN n UNION ALL MATCH(n) RETURN n$$) AS (result vertex);
ERROR:  could not identify an equality operator for type vertex
LINE 2: ... * FROM cypher('cypher_union', $$MATCH (n) RETURN n UNION MA...
                                                             ^
/*should return 3 null rows*/
SELECT * FROM cypher('cypher_union', $$RETURN NULL UNION ALL RETURN NULL UNION ALL RETURN NULL$$) AS (result gtype);
 result 
--------
 
 
 
(3 rows)

/*should return 1 null row*/
SELECT * FROM cypher('cypher_union', $$RETURN NULL UNION ALL RETURN NULL UNION RETURN NULL$$) AS (result gtype);
 result 
--------
 
(1 row)

/*should return 2 null rows*/
SELECT * FROM cypher('cypher_union', $$RETURN NULL UNION RETURN NULL UNION ALL RETURN NULL$$) AS (result gtype);
 result 
--------
 
 
(2 rows)

/* scoping */
SELECT * FROM cypher('cypher_union', $$MATCH (n) RETURN n UNION ALL MATCH (m) RETURN n$$) AS (result vertex);
ERROR:  could not find rte for n
LINE 2: ..., $$MATCH (n) RETURN n UNION ALL MATCH (m) RETURN n$$) AS (r...
                                                             ^
/*
 *UNION and UNION ALL, type casting
 */
SELECT * FROM cypher('cypher_union', $$RETURN 1.0::int UNION RETURN 1::float UNION ALL RETURN 2.0::float$$) AS (result gtype);
 result 
--------
 1
 2.0
(2 rows)

SELECT * FROM cypher('cypher_union', $$RETURN 1.0::int UNION RETURN 1.0::float UNION ALL RETURN 1::int$$) AS (result gtype);
 result 
--------
 1
 1
(2 rows)

SELECT * FROM cypher('cypher_union', $$RETURN 1.0::float UNION RETURN 1::int UNION RETURN 1::float$$) AS (result gtype);
 result 
--------
 1.0
(1 row)

SELECT drop_graph('cypher_union', true);
NOTICE:  drop cascades to 2 other objects
DETAIL:  drop cascades to table cypher_union._ag_label_vertex
drop cascades to table cypher_union._ag_label_edge
NOTICE:  graph "cypher_union" has been dropped
 drop_graph 
------------
 
(1 row)

